File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    q = deque()
    dx = [1,0,-1,0]
    dy = [0,-1,0,1]
    m = len(maps)
    n = len(maps[0])
    q.append((0,0))
    visit = [[0] * (n) for _ in range(m)]
    
    while q:
        x,y = q.popleft()
        for i in range(4):
            nx=x+dx[i]
            ny=y+dy[i]
            if 0 <= nx < m and 0 <= ny < n and maps[nx][ny] == 1 and visit[nx][ny] == 0:
                visit[x+dx[i]][y+dy[i]] = 1
                q.append((x+dx[i],y+dy[i]))
                maps[x+dx[i]][y+dy[i]] = maps[x][y] + 1
                
    if maps[m-1][n-1] == 1:
        return -1
    else:
        return maps[m-1][n-1]
    # return maps
    # return dfs(maps,1,1,m,n,visit)
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        int[][] dir = new int[][]{{0,1},{1,0},{0,-1},{-1,0}};
        int row = maps.length;
        int col = maps[0].length;
        int answer = -1;
        
        Deque<int[]> q = new ArrayDeque<>();
        q.add(new int[] {0,0});
        while(!q.isEmpty()){
            int[] out = q.remove();
            int y = out[0], x = out[1];
            if(y == row-1 && x == col-1){
                answer = maps[y][x];
                break;
            }
            for(int[] d : dir) {
                int ny = y + d[0], nx = x + d[1];
                if(0 <= ny && ny < row && 0 <= nx && nx < col && maps[ny][nx] == 1){
                    maps[ny][nx] = maps[y][x] + 1;
                    q.add(new int[] {ny,nx});
                }
            }
        }
        return answer;
    }
}